subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -275,87 +328,291 @@ func (h *HivedAlgorithm) GetAffinityGroup(name string) api.AffinityGroup {
+			bindCell(physical, virtual)
+		}
+	}
+}
+
+// initBadNodes marks all the physical nodes defined in the config as bad,
+// and wait for K8s's AddNode calls to recognize the healthy nodes.
+func (h *HivedAlgorithm) initBadNodes() {
+	klog.Info("Setting all nodes defined in the config to bad first, " +
+		"and wait for K8s to recognize the healthy nodes (addNode)")
+	for _, ccl := range h.fullCellList {
+		for _, c := range ccl[CellLevel(len(ccl))] {
+			nodes, _ := c.(*PhysicalCell).GetPhysicalPlacement()
+			for _, n := range nodes {
+				h.setBadNode(n)
+			}
+		}
+	}
+}
+
+// setBadNode marks a node and the cells in it as bad.
+func (h *HivedAlgorithm) setBadNode(nodeName string) {
+	if h.badNodes.Contains(nodeName) {
+		return
+	}
+	h.badNodes.Add(nodeName)
+	for _, ccl := range h.fullCellList {
+		for _, gpu := range ccl[1] {
+			pGpu := gpu.(*PhysicalCell)
+			nodes, _ := pGpu.GetPhysicalPlacement()
+			if nodes[0] == nodeName {
+				h.setBadCell(pGpu)
+			}
+		}
+	}
+}
+
+// setBadNode marks a node and the cells in it as healthy.
+func (h *HivedAlgorithm) setHealthyNode(nodeName string) {
+	if !h.badNodes.Contains(nodeName) {
+		return
+	}
+	h.badNodes.Delete(nodeName)
+	for _, ccl := range h.fullCellList {
+		for _, gpu := range ccl[1] {
+			pGpu := gpu.(*PhysicalCell)
+			nodes, _ := pGpu.GetPhysicalPlacement()
+			if nodes[0] == nodeName {
+				h.setHealthyCell(pGpu)
+			}
+		}
+	}
+}
+
+// setBadCell marks a physical cell as bad and recursively for its parent, guaranteeing that
+// a cell is bad if all of its children are bad. The virtual cell that this physical cell
+// is bound to will also be marked as bad. If the cell has not been bound to a virtual cell,
+// but the scheduler finds that it causes a VC's free cell doomed to be bad,
+// one of the cell in that VC will be marked as bad (the first free cell of this type in the VC).
+func (h *HivedAlgorithm) setBadCell(c *PhysicalCell) {
+	klog.Infof("Cell %v is set to bad", c.GetAddress())
+	c.GetAPIStatus().CellHealthiness = api.CellBad
+	if c.virtualCell != nil {
+		c.GetAPIStatus().VirtualCell.CellHealthiness = api.CellBad
+		c.virtualCell.GetAPIStatus().CellHealthiness = api.CellBad
+		c.virtualCell.GetAPIStatus().PhysicalCell.CellHealthiness = api.CellBad
+	}
+	chain, level := c.GetChain(), c.GetLevel()
+	if inFreeCellList(c) {
+		h.badFreeCellNum[chain][level]++
+		if h.allVCFreeCellNum[chain][level] > h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level] {
+			klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (healthy %v, bad %v) "+
+				"than the total free cells of all the VCs (%v). Certain VCs' cells may be doomed to be bad.",
+				h.cellTypes[chain][level], chain, level, h.totalLeftCellNum[chain][level]-h.badFreeCellNum[chain][level],
+				h.badFreeCellNum[chain][level], h.allVCFreeCellNum[chain][level])
+			for vcName, vcQuota := range h.vcFreeCellNum {
```

Comment:
>vcQuota [](start = 15, length = 7)

Rename quota to be a more concert concept. Such as total/free cells. Also for all places.